File: back-bardak/src/main/java/kz/bardak/game/rules/RulesConfig.java
package kz.bardak.game.rules;

import java.util.Objects;

/**
 * Игровые числа стола. Явное требование §1.6: движок не опирается на конкретные числа —
 * в коде не должно быть литералов {@code 6}, {@code 5}, {@code 30}. Всё живёт здесь,
 * попадает в {@code rules_snapshot} матча и параметризует тесты.
 *
 * @param dealSize            сколько карт на руках после добора
 * @param maxAttackFirstRound потолок атакующих карт, пока в отбое пусто (§1.5)
 * @param maxAttackPerRound   потолок атакующих карт после первой биты — он же потолок защиты
 * @param transfersEnabled    разрешены ли переводы (§2.2)
 * @param jokersEnabled       участвуют ли джокеры в колоде
 * @param navesEnabled        включена ли центральная механика — навесы (§2.3)
 * @param navesScale          длина личной шкалы; укоротить её должно быть правкой конфига
 */
public record RulesConfig(
        int dealSize,
        int maxAttackFirstRound,
        int maxAttackPerRound,
        boolean transfersEnabled,
        boolean jokersEnabled,
        boolean navesEnabled,
        NavesScale navesScale) {

    public RulesConfig {
        requirePositive(dealSize, "dealSize");
        requirePositive(maxAttackFirstRound, "maxAttackFirstRound");
        requirePositive(maxAttackPerRound, "maxAttackPerRound");
        Objects.requireNonNull(navesScale, "navesScale");
    }

    /** Значения по умолчанию из §1.6 — стартовая точка стола, не константы движка. */
    public static RulesConfig defaults() {
        return new RulesConfig(6, 5, 6, true, true, true, NavesScale.full());
    }

    /** Каркас без навесов: серия раздач без продвижения по шкале, годится только для отладки. */
    public RulesConfig withoutNaves() {
        return new RulesConfig(dealSize, maxAttackFirstRound, maxAttackPerRound,
                transfersEnabled, jokersEnabled, false, navesScale);
    }

    public RulesConfig withTransfersDisabled() {
        return new RulesConfig(dealSize, maxAttackFirstRound, maxAttackPerRound,
                false, jokersEnabled, navesEnabled, navesScale);
    }

    /**
     * Потолок атакующих карт в текущем раунде. Зависит не от того, что лежит на столе,
     * а от того, уходили ли уже карты в отбой в этой раздаче (§1.5, ADR-023).
     */
    public int attackLimit(final boolean anyPileDiscarded) {
        return anyPileDiscarded ? maxAttackPerRound : maxAttackFirstRound;
    }

    private static void requirePositive(final int value, final String name) {
        if (value <= 0) {
            throw new IllegalArgumentException("%s должен быть положительным, получено: %d"
                    .formatted(Objects.requireNonNull(name, "name"), value));
        }
    }
}

File: back-go/internal/domain/game/rules_config.go
package game

import "fmt"

// NoNaves — уровень игрока, которому ещё ничего не навешивали.
const NoNaves = -1

// NavesScale — личная шкала навесов: 6 → 7 → … → A → Joker.
//
// ⭐ Счёта в очках в игре нет — шкала и есть счёт.
//
// Уровень игрока кодируется индексом:
//
//	-1            навесов ещё не было, летит первая ступень
//	0 … size-1    навешен Ranks[level]
//	size          навешен джокер — игрок проиграл
//
// Длина шкалы — параметр стола: укоротить её до 9…A должно быть правкой конфигурации,
// а не кода.
type NavesScale struct {
	// Ranks — ступени по возрастанию; джокер — терминальная ступень сверх списка.
	Ranks []Rank
}

// NewNavesScale собирает шкалу. Пустая шкала бессмысленна: навешивать было бы нечего.
func NewNavesScale(ranks []Rank) (NavesScale, error) {
	if len(ranks) == 0 {
		return NavesScale{}, fmt.Errorf("шкала навесов не может быть пустой")
	}
	out := make([]Rank, len(ranks))
	copy(out, ranks)
	return NavesScale{Ranks: out}, nil
}

// FullNavesScale — полная шкала из всех девяти рангов, как играют вживую.
func FullNavesScale() NavesScale {
	return NavesScale{Ranks: Ranks()}
}

// JokerLevel — уровень, после которого следующая ступень джокер.
func (s NavesScale) JokerLevel() int { return len(s.Ranks) }

// IsFinished — джокер уже навешен: игрок проиграл, навешивать ему больше нечего.
func (s NavesScale) IsFinished(level int) bool { return level >= s.JokerLevel() }

// NextIsJoker — следующая ступень джокер, а не обычный ранг.
func (s NavesScale) NextIsJoker(level int) bool { return level+1 == s.JokerLevel() }

// NextRank — ранг, который сейчас «летит» игроку.
//
// Второе значение false, если следующая ступень джокер либо шкала уже пройдена.
func (s NavesScale) NextRank(level int) (Rank, bool) {
	if level+1 >= s.JokerLevel() || level+1 < 0 {
		return 0, false
	}
	return s.Ranks[level+1], true
}

// IsFlyingCard — подходит ли карта под текущую ступень жертвы. Масть значения не имеет.
func (s NavesScale) IsFlyingCard(level int, card Card) bool {
	if s.IsFinished(level) {
		return false
	}
	if s.NextIsJoker(level) {
		_, isJoker := card.(Joker)
		return isJoker
	}
	rank, ok := s.NextRank(level)
	if !ok {
		return false
	}
	pip, isPip := card.(Pip)
	return isPip && pip.Rank == rank
}

// RulesConfig — правила стола.
//
// ⭐ Все игровые числа живут здесь, а не литералами в движке: «подкидывать не больше
// шести» — это правило стола, и менять его должна настройка, а не правка кода.
type RulesConfig struct {
	DealSize            int
	MaxAttackFirstRound int
	MaxAttackPerRound   int
	TransfersEnabled    bool
	JokersEnabled       bool
	NavesEnabled        bool
	NavesScale          NavesScale
}

// DefaultRulesConfig — стартовая точка стола, а не константы движка.
func DefaultRulesConfig() RulesConfig {
	return RulesConfig{
		DealSize:            6,
		MaxAttackFirstRound: 5,
		MaxAttackPerRound:   6,
		TransfersEnabled:    true,
		JokersEnabled:       true,
		NavesEnabled:        true,
		NavesScale:          FullNavesScale(),
	}
}

// WithoutNaves — каркас без навесов: серия раздач без продвижения по шкале.
// Годится только для отладки.
func (c RulesConfig) WithoutNaves() RulesConfig {
	next := c
	next.NavesEnabled = false
	return next
}

// AttackLimit — потолок атакующих карт в раунде.
//
// ⚠️ Зависит от того, уходили ли в этой РАЗДАЧЕ карты в отбой, а не от состояния стола:
// пять до первого отбоя, шесть после.
func (c RulesConfig) AttackLimit(anyPileDiscarded bool) int {
	if anyPileDiscarded {
		return c.MaxAttackPerRound
	}
	return c.MaxAttackFirstRound
}

// Validate — проверка на осмысленность; вызывается при разборе настроек стола.
func (c RulesConfig) Validate() error {
	for _, check := range []struct {
		value int
		name  string
	}{
		{c.DealSize, "dealSize"},
		{c.MaxAttackFirstRound, "maxAttackFirstRound"},
		{c.MaxAttackPerRound, "maxAttackPerRound"},
	} {
		if check.value <= 0 {
			return fmt.Errorf("%s должен быть положительным, получено: %d", check.name, check.value)
		}
	}
	if len(c.NavesScale.Ranks) == 0 {
		return fmt.Errorf("шкала навесов не может быть пустой")
	}
	return nil
}
